xplique/concepts/craft.py

```diff
@@ -0,0 +1,604 @@
+                       batch_size: int = 64,
+                       patch_size: int = 64):
+        super().__init__(number_of_concepts, batch_size)
+        self.input_to_latent_model = input_to_latent_model
+        self.latent_to_logit_model = latent_to_logit_model
+        self.patch_size = patch_size
+        self.cmaps = None
+
+        # sanity checks
+        assert(callable(input_to_latent_model)), "input_to_latent_model must be a callable function"
+        assert(callable(latent_to_logit_model)), "latent_to_logit_model must be a callable function"
+
+    @abstractmethod
+    def _latent_predict(self, inputs: np.ndarray):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _logit_predict(self, inputs: np.ndarray):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _extract_patches(self, inputs: np.ndarray):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _to_np_array(self, input, dtype) -> np.ndarray:
+        raise NotImplementedError
+
+    def fit(self, inputs : np.ndarray, class_id: int) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Fit the Craft model to the input data.
+
+        Parameters
+        ----------
+        inputs
+            Input data of shape (n_samples, height, width, channels).
+            (x1, x2, ..., xn) in the paper.
+        class_id
+            The class id of the inputs.
+
+        Returns
+        -------
+        (X, U, W)
+            A tuple containing the crops (X in the paper),
+            the concepts values (U) and the concepts basis (W).
+        """
+        # extract patches from the input data
+        self.inputs = inputs
+        self.class_id = class_id
+        self.crops, activations = self._extract_patches(self.inputs)
+
+        # apply NMF to the activations to obtain matrices U and W
+        reducer = NMF(n_components=self.number_of_concepts, alpha_W=1e-2)
+        self.crops_u = reducer.fit_transform(activations)
+        self.W = reducer.components_.astype(np.float32)
+
+        # store the factorizer as attribute of the Craft instance
+        self.reducer = reducer
+
+        return self.crops, self.crops_u, self.W
+
+    def transform(self, inputs : np.ndarray, activations : np.ndarray = None):
+        """Transforms the inputs data into its concept representation.
+
+        Parameters
+        ----------
+        inputs : numpy array or Tensor
+            The input data to be transformed.
+        activations
+            Pre-computed activations of the input data. If not provided, the activations
+            will be computed using the input_to_latent_model model on the inputs.
+
+        Returns
+        -------
+        U
+            The concept value (U) of the inputs.
+        """
+        self.check_if_fitted()
+
+        if activations == None:
+            activations = self._latent_predict(inputs)
+
+        is_4d = len(activations.shape) == 4
```

Comment:
Has it been tested in other cases? Otherwise, I suggest a `assert len(activations.shape) == 4`. Then the if conditions can be removed.

Reply:
Yes, 4 dims and 2 dims are supported and tested in the unit tests.